pkg/virt-controller/watch/vmi.go

```diff
@@ -433,48 +432,42 @@ func (c *VMIController) syncDynamicLabelsToPod(vmi *virtv1.VirtualMachineInstanc
 		return nil
 	}
 
-	newLabelBytes, err := json.Marshal(podMeta.Labels)
-	if err != nil {
-		return err
-	}
 	if pod.ObjectMeta.Labels == nil {
-		patchOps = append(patchOps, fmt.Sprintf(`{ "op": "add", "path": "/metadata/labels", "value": %s }`, string(newLabelBytes)))
+		patches.AddAddOp("/metadata/labels", podMeta.Labels)
 	} else {
-		oldLabelBytes, err := json.Marshal(pod.ObjectMeta.Labels)
-		if err != nil {
-			return err
-		}
-		patchOps = append(patchOps, fmt.Sprintf(`{ "op": "test", "path": "/metadata/labels", "value": %s }`, string(oldLabelBytes)))
-		patchOps = append(patchOps, fmt.Sprintf(`{ "op": "replace", "path": "/metadata/labels", "value": %s }`, string(newLabelBytes)))
+		patches.AddTestOp("/metadata/labels", pod.ObjectMeta.Labels)
+		patches.AddReplaceOp("/metadata/labels", podMeta.Labels)
 	}
 
-	patchBytes := controller.GeneratePatchBytes(patchOps)
-
-	if len(patchBytes) == 0 {
+	if patches.IsEmpty() {
 		return nil
 	}
+	patch, err := patches.GeneratePatchPayload()
+	if err != nil {
+		return err
+	}
 
-	if _, err := c.clientset.CoreV1().Pods(pod.Namespace).Patch(context.Background(), pod.Name, types.JSONPatchType, patchBytes, v1.PatchOptions{}); err != nil {
+	if _, err := c.clientset.CoreV1().Pods(pod.Namespace).Patch(context.Background(), pod.Name, types.JSONPatchType, patch, v1.PatchOptions{}); err != nil {
 		log.Log.Object(pod).Errorf("failed to sync dynamic pod labels during sync: %v", err)
 		return err
 	}
 	return nil
 }
 
 func (c *VMIController) syncPodAnnotations(pod *k8sv1.Pod, newAnnotations map[string]string) (*k8sv1.Pod, error) {
-	var patchOps []string
+	patches := patch.NewPatchSet()
 	for key, newValue := range newAnnotations {
 		if podAnnotationValue, keyExist := pod.Annotations[key]; !keyExist || (keyExist && podAnnotationValue != newValue) {
-			patchOp, err := prepareAnnotationsPatchAddOp(key, newValue)
-			if err != nil {
-				return nil, err
-			}
-			patchOps = append(patchOps, patchOp)
+			key = patch.EscapeJSONPointer(key)
+			patches.AddAddOp(fmt.Sprintf("/metadata/annotations/%s", key), newValue)
 		}
 	}
 	var patchedPod *k8sv1.Pod
-	patchBytes := controller.GeneratePatchBytes(patchOps)
-	if len(patchBytes) > 0 {
+	patchBytes, err := patches.GeneratePatchPayload()
```

Comment:
super tiny nit: in line 445 you refer to this as "patch"  but here you use patchBytes for what appears to be the same thing. It might be good to make this consistent across both?